daemon/daemon.go

```diff
@@ -1157,14 +1190,121 @@ func NewDaemon(ctx context.Context, config *config.Config, pluginStore *plugin.S
 			}
 		}
 
-		// TODO: imageStore, distributionMetadataStore, and ReferenceStore are only
-		// used above to run migration. They could be initialized in ImageService
-		// if migration is called from daemon/images. layerStore might move as well.
-		d.imageService = images.NewImageService(imgSvcConfig)
+		// If no containers are running, check whether can migrate image service
+		if drv := layerStore.DriverName(); len(containers[drv]) == 0 && migrationThreshold >= 0 {
+			switch drv {
+			case "overlay2":
+				driverName = "overlayfs"
+			case "windowsfilter":
+				driverName = "windows"
+			case "vfs":
+				driverName = "native"
+			default:
+				migrationThreshold = -1
+				log.G(ctx).Infof("Not migrating to containerd snapshotter, no migration defined for graph driver %q", drv)
+			}
+
+			var totalSize int64
+			ic := imgSvcConfig.ImageStore.Len()
+			if migrationThreshold >= 0 && ic > 0 {
+				for _, img := range imgSvcConfig.ImageStore.Map() {
+					if layerID := img.RootFS.ChainID(); layerID != "" {
+						l, err := imgSvcConfig.LayerStore.Get(layerID)
+						if err != nil {
+							if errors.Is(err, layer.ErrLayerDoesNotExist) {
+								continue
+							}
+							return nil, err
+						}
+
+						// Just look at layer sizze for considering maximum size
+						totalSize += l.Size()
+						layer.ReleaseAndLog(imgSvcConfig.LayerStore, l)
+					}
+				}
+
+			}
+
+			if totalSize <= migrationThreshold {
```

Comment:
https://github.com/moby/moby/blob/3023608e18e86df5d0e5a48def4fc0fe86992df4/daemon/daemon.go#L877

Looks like this (`totalSize <= math.MaxInt64`) would always be true regardless if there are images left or not?

Reply:
Can we simplify this whole `migrationThreshold` to just a boolean `doMigrate := true/false` to simplify this for now, especially since we don't support setting the threshold?

Reply:
This value can also be set to 0 though, indicating that we should only switch to containerd when image count goes to zero.

Reply:
Right but it's only for windows in one specific case:

https://github.com/moby/moby/blob/3023608e18e86df5d0e5a48def4fc0fe86992df4/daemon/daemon.go#L1088-L1103

I think this should default to `0` so that we only migrate once users reach empty state.

https://github.com/dmcgowan/docker/blob/8f7bd1f547b13ef5fbfa5b5d9f5a3f3b287bf67e/daemon/daemon.go#L878

Reply:
The migration will need more thought and testing before no longer being experimental. Changing the default threshold to 0 would change the logic though and would necessitate adding the threshold in anyway in order to test it. Can add the threshold in a follow up, it is trivial, but it is experimental and not the most important aspect of this PR.